README.rst
``repoze.postoffice`` README
============================

``repoze.postoffice`` provides a centralized depot for collecting incoming
email for consumption by multiple applications.  Incoming mail is sorted
into queues according to rules with the expectation that each application
will then consume its own queue.  Each queue is a first-in-first-out (FIFO)
queue, so messages are processed in the order received.

ZODB is used for storage and is also used to provide the client interface.
``repoze.postoffice`` clients create a ZODB connection and manipulate models.
This makes consuming the message queue in the context of a transaction,
relatively simple.

Please see ``docs/index.rst`` for the complete documentation.


CHANGES.rst
``repoze.postoffice`` Changelog
===============================

0.25 (2014-09-30)
-----------------

- Introduce the 'find_queue' function as an alternative to 'open_queue' for 
  applications that will manage their own database connection.

0.24 (2013-10-10)
-----------------

- Avoid throttling BCC'ed messages (those where ``X-Original-To`` is not
  present in ``To`` or ``CC``), because they can't be part of loops.

- Accept messages as non-duplicates which have the same ``Message-ID`` headers,
  but different ``X-Original-To`` headers.  Rationale:  on systems which
  add ``X-Original-To`` (postfix adds it by default), its presence indicates
  delivery to a separate user; we should allow it to be queued for attempted
  processing.

0.23 (2013-07-31)
-----------------

- Enable configuration of the headers considred by the ``to_hostname`` filter.

- Update the headers considered by in the ``to_hostname`` filter to
  include ``X-Original-To`` by default.  This header is set by Postfix
  to match the envelope recipient;  including it allows matching messages
  for which the domain is BCC'ed or delivered from a mailing list.

0.22 (2013-07-26)
-----------------

- Avoid erroring while logging messages with encoded header values
  E.g., some clients don't escape non-ASCII in ``To:`` / ``From:`` /
  ``Subject:`` headers.  See LP #1202770.

- Handle broken addresses in ToHostName filter more gracefully.

- Converted documentation to Sphinx.

- Moved repository to Github.

0.21 (2012-09-05)
-----------------

- Make regexp filters case sensitive.

0.20 (2012-05-14)
-----------------

- Only close databases we create ourselves in open_queue.

0.19 (2012-05-10)
-----------------

- Allow calling application to provide an already open database to open_queue.
  (LP #985546)

0.18 (2012-04-19)
-----------------

- Added graceful degeneration when provided with uknown character encoding in 
  a MIME text part.

0.17 (2011-09-26)
-----------------

- Added a header, 'X-Postoffice-Date', to queued messages.  It records
  the time each message was received (as seconds since the epoch.)  It
  is set to the modified time of the maildir message file.  

0.16 (2011-06-30)
-----------------

- Added better fault tolerance for insane date headers generated by spambots.
  (LP #697033)

0.15 (2011-06-15)
-----------------

- Body checks are now multiline regexp checks. (LP #787573)

0.14 (2011-05-17)
-----------------

- Fixed problem where the zodb_uri could be unicode, which eventually breaks
  the ZEO client for ZEO uris.  zodb_uri is now converted to a UTF-8 string.

0.13 (2011-05-05)
-----------------

- Fixed problem with header filters not working properly with non-ASCII
  characters in headers.  (LP #777455)

- Fixed bug in regular expression body filter which improperly parsed the
  'Content-Type' header in order to extract the character set of a Mime part.

0.12 (2011-04-25)
-----------------

- Respect leading and trailing whitespace in rules files.

0.11 (2011-04-25)
-----------------

- When a message is rejected by a filter, a message is logged showing which
  filter triggered the rejection.

0.10 (2011-04-20)
-----------------

- Improved logging output now includes a timestamp.

- Worked around (probable) bug in stdlib email parser where a message part
  might have a charset set in the 'Content-Type' header, but
  message.get_charset() returns None.

0.9 (2011-04-15)
----------------

- Added greater fault tolerance for malformed email addresses.  (Shakes fist at
  spammers.)

- Added four new filter types based on regular expression matching:
  ``header_regexp``, ``header_regexp_file``, ``body_regexp``,
  ``body_regexp_file``.  See README.txt for information on how to use these
  new filters.

- Added a new option to the global configuration: ``reject_filters``.
  This allows you to set up filters at a global level for rejecting certain
  messages.  See README.txt for more information.

0.8 (2011-01-14)
----------------

- The 'to_hostname' filter now parses multiple email addresses and checks the
  'Cc' header as well as the 'To' header.  (LP #659243)

- If multiple incoming messages in a 24 hour period have the same Message-Id,
  they are presumed to be duplicates and all but the first are discarded.
  (LP #659243)

0.7 (2010-09-15)
----------------

- Fixed another case where non-RFC 2047 compliant headers could cause an
  exception to be raised.  (LP #637484)

0.6 (2010-09-13)
----------------

- Added Queue.requeue_quarantined_messages() convenience method to API.

- Allow for multiple hosts in 'to_hostname' filter. (LP #614528)

- Added graceful degradation for non-RFC 2047 compliant headers, in order to
  avoid crashing when spambots send us malformed messages. (LP #637484)

0.5 (2010-08-03)
----------------

- Added 'X-Postoffice: Bounced' header to outgoing bounce and quarantine
  messages. The presence of this header is checked when importing messages and
  any messages which contain it are discarded. This is to prevent possible
  ricochets of bounce messages back into the system. (LP #612587)

- Incoming messages with a 'From' header which matches exactly its 'To' header
  are now discarded as probable spam. (LP #612588)

0.4 (2010-07-30)
----------------

- Fixed bug in processing body of bounce messages when non-ascii unicode
  characters are present.

0.3 (2010-07-20)
----------------

- Fixed divide by zero error when calculating instantaneous message frequency.

- Fixed bug in repoze.postoffice.queue.open_queue where a ZEO connection would
  be left open if there was a KeyError on the queue name.


repoze/postoffice/script.py
from __future__ import with_statement

from code import interact
from optparse import OptionParser
from repoze.postoffice.api import PostOffice
import logging
import os
import sys

logging.basicConfig(
    format='%(asctime)s %(levelname)s %(name)s: %(message)s')


class ConsoleScript(object):
    """
    Populates postoffice queues from a Maildir.
    """

    def __init__(self, argv=sys.argv[1:], find_config=None):
        parser = OptionParser(description=self.__doc__)
        parser.add_option('-C', '--config', dest='config', default=None,
                          help='Path to configuration ini file.',
                          metavar='FILE')
        parser.add_option('-v', '--verbose', dest='verbose', default=False,
                          action='store_true',
                          help='Print info level log messages')

        options, args = parser.parse_args(argv)
        if args:
            parser.error('Extra arguments given.')

        config = options.config
        if config is None:
            config = _find_config()
        if config is None:
            parser.error('Unable to find configuration file.')

        log_level = logging.WARN
        if options.verbose:
            log_level = logging.INFO
        self.log = logging.getLogger('repoze.postoffice')
        self.log.setLevel(log_level)
        self.config = config

    def __call__(self):
        po = PostOffice(self.config)
        po.reconcile_queues(self.log)
        po.import_messages(self.log)

    def debug(self):
        po = PostOffice(self.config)
        banner = '"root" is the root queues folder.'
        with po._get_root() as root:
            interact(banner, local={'root':root})

def _find_config():
    path = os.path.abspath('postoffice.ini')
    if os.path.exists(path):
        return path

    path = os.path.join(os.path.abspath('etc'), 'postoffice.ini')
    if os.path.exists(path):
        return path

    base = os.path.dirname(os.path.dirname(os.path.abspath(sys.argv[0])))
    path = os.path.join(base, 'etc', 'postoffice.ini')
    if os.path.exists(path):
        return path

    path = '/etc/postoffice.ini'
    if os.path.exists(path):
        return path

def main():
    return ConsoleScript()()

def debug():
    return ConsoleScript().debug()

if __name__ == '__main__':
    main()


repoze/postoffice/filters.py
from __future__ import with_statement

import codecs
import re

from repoze.postoffice.message import decode_header

_STANDARD_TO_HEADERS = ('To', 'Cc', 'X-Original-To')

class ToHostnameFilter(object):
    """Test the hostname of the email address in the specified message headers.
    """
    def __init__(self, expr, headers=_STANDARD_TO_HEADERS):
        if ';' in expr:
            expr, attrs = expr.split(';', 1)
            for a_expr in attrs.split(';'):
                name, value = [x.strip() for x in a_expr.split('=')]
                if name == 'headers':
                    headers = [x.strip() for x in value.split(',')]
                else:
                    raise ValueError('Unknown config attribute: %s' % name)
        self.expr = expr
        self.domains = expr.lower().split()
        self.headers = headers

    def __call__(self, message):
        addrs = []
        for header in self.headers:
            value = message.get(header)
            if not value:
                continue
            addrs.extend(value.split(','))

        for addr in addrs:
            lt = addr.find('<')
            if lt != -1:
                gt = addr.rfind('>')
                if gt == -1:
                    gt = None
                addr = addr[lt+1:gt]
            if '@' not in addr:
                continue
            hostname = addr.split('@')[1].lower()

            for domain in self.domains:
                if domain.startswith('.') and hostname.endswith(domain[1:]):
                    return 'to_hostname: %s matches %s' % (addr, domain)
                if hostname == domain:
                    return 'to_hostname: %s matches %s' % (addr, domain)

        return None


class HeaderRegexpFilter(object):
    """
    Matches a regular expression on the headers of an email message.
    """
    def __init__(self, *exprs):
        self.regexps = [(expr, re.compile(expr))
                        for expr in exprs]

    def __call__(self, message):
        for name in message.keys():
            header = '%s: %s' % (name, decode_header(message.get(name)))
            for regexp, compiled in self.regexps:
                if compiled.match(header) is not None:
                    return 'header_regexp: headers match %s' % repr(regexp)
        return None


class HeaderRegexpFileFilter(HeaderRegexpFilter):
    """
    Same as HeaderRegexpFilter but loads regexps from a file.
    """
    def __init__(self, path):
        self.regexps = regexps = []
        with codecs.open(path, 'r', 'UTF-8') as f:
            for line in f:
                expr = line.rstrip('\n').rstrip('\r')
                regexps.append((expr, re.compile(expr)))


class BodyRegexpFilter(object):
    """
    Matches a regular expression on the body of an email message (any part).
    """
    def __init__(self, *exprs):
        self.regexps = [(expr, re.compile(expr, re.MULTILINE))
                        for expr in exprs]

    def __call__(self, message):
        for part in message.walk():
            if not part.get_content_type().startswith('text/'):
                continue

            # Get body for this message part as unicode
            body = part.get_payload(decode=True)
            charset = part.get_charset()
            if charset is None:
                content_type = part.get('Content-Type')
                if content_type is not None and 'charset=' in content_type:
                    double_slash = content_type.find('//')
                    if double_slash != -1:
                        content_type = content_type[:double_slash].strip()
                    for piece in content_type.split(';'):
                        piece = piece.strip()
                        if piece.startswith('charset='):
                            charset = piece[8:]
                            break
            else:
                charset = str(charset)

            try_charsets = filter(None, [charset, 'UTF-8', 'ISO-8859-1'])
            for charset in try_charsets:
                try:
                    body = body.decode(charset)
                    break
                except (LookupError, UnicodeError):
                    pass

            # See if we match
            for regexp, compiled in self.regexps:
                if compiled.search(body) is not None:
                    return 'body_regexp: body matches %s' % repr(regexp)

        return None

class BodyRegexpFileFilter(BodyRegexpFilter):
    """
    Same as BodyRegexpFilter but loads regexps from a file.
    """
    def __init__(self, path):
        self.regexps = regexps = []
        with codecs.open(path, 'r', 'UTF-8') as f:
            for line in f:
                expr = line.rstrip('\n').rstrip('\r')
                regexps.append((expr, re.compile(expr, re.MULTILINE)))


repoze/postoffice/api.py
from __future__ import with_statement

import codecs
from cStringIO import StringIO
from ConfigParser import ConfigParser
from contextlib import contextmanager
import datetime
from email.utils import parsedate
from mailbox import Maildir
from mailbox import MaildirMessage
from mailbox import NoSuchMailboxError
import os
import re
import shutil
import smtplib
import transaction

from repoze.postoffice import filters
from repoze.postoffice.queue import QueuesFolder
from repoze.postoffice.queue import Queue
from repoze.zodbconn.uri import db_from_uri

filter_factories = {
    'to_hostname': filters.ToHostnameFilter,
    'header_regexp': filters.HeaderRegexpFilter,
    'header_regexp_file': filters.HeaderRegexpFileFilter,
    'body_regexp': filters.BodyRegexpFilter,
    'body_regexp_file': filters.BodyRegexpFileFilter,
}

MAIN_SECTION = 'post office'
_marker = object()

try:
    unicode
except NameError: #pragma NO COVER Py3k
    _TEXT = str
else: #pragma NO COVER Python2
    _TEXT = unicode


class PostOffice(object):
    """
    Provides server side API for repoze.postoffice.
    """

    # Overridable for testing
    Maildir = Maildir
    MaildirMessage = MaildirMessage
    Queue = Queue

    def __init__(self, filename, db_from_uri=db_from_uri, open=open):
        """
        Initialize from configuration file.
        """
        # db_from_uri and open passed in for unittesting
        here = os.path.dirname(os.path.abspath(filename))
        codec = codecs.lookup('UTF-8')
        fp = codec.streamreader(_load_fp(open(filename)))
        self._section_indices = _get_section_indices(fp)
        fp.seek(0)
        config = ConfigParser(defaults=dict(here=here))
        config.readfp(fp)
        self._init_main_section(config)
        self._get_root = _RootContextManagerFactory(
            self.zodb_uri, db_from_uri, self.zodb_path
        )
        self._init_queues(config)

    def _init_main_section(self, config):
        if not config.has_section(MAIN_SECTION):
            raise ValueError('Config file is missing required section: %s' %
                             MAIN_SECTION)

        self.zodb_uri = _get_opt(config, MAIN_SECTION, 'zodb_uri')
        if isinstance(self.zodb_uri, unicode):
            self.zodb_uri = self.zodb_uri.encode('UTF-8')
        self.maildir = _get_opt(config, MAIN_SECTION, 'maildir')
        self.zodb_path = _get_opt(config, MAIN_SECTION, 'zodb_path',
                                     '/postoffice')
        self.ooo_loop_frequency = _get_opt_float(
            config, MAIN_SECTION, 'ooo_loop_frequency', '0')
        self.ooo_loop_headers = _get_opt_list(
            config, MAIN_SECTION, 'ooo_loop_headers', '')
        self.ooo_throttle_period = datetime.timedelta(seconds=_get_opt_int(
            config, MAIN_SECTION, 'ooo_throttle_period', '300'))
        self.max_message_size = _get_opt_bytes(
            config, MAIN_SECTION, 'max_message_size', '0')

        self.reject_filters = filters = []
        filters_setting = _get_opt(config, MAIN_SECTION, 'reject_filters', None)
        if filters_setting is not None:
            for filter in [f.strip() for f in
                           filters_setting.strip().split('\n')]:
                filters.append(self._init_filter(filter))


    def _init_queues(self, config):
        queues = []
        for section in config.sections():
            if section.startswith('queue:'):
                queues.append(self._init_queue(config, section))

        # ConfigParser doesn't preserve section ordering, but we're interested
        # processing queues in the order they appear in the config file, so
        # we've done our own scan to pick out the sections and their relative
        # indices and use that now to sort the queues.
        section_indices = self._section_indices
        queues.sort(key=lambda q: section_indices[q['section']])
        self.configured_queues = queues

    def _init_queue(self, config, section):
        name = section[6:] # len('queue:') == 6
        filters = []
        for option in config.options(section):
            if option == 'filters':
                for filter_ in [f.strip() for f in
                                config.get(section, option)
                                .strip().split('\n')]:
                    filters.append(self._init_filter(filter_))
            elif option == 'here':
                pass
            else:
                raise ValueError('Unknown config parameter for queue: %s' %
                                 option)

        return dict(name=name, filters=filters, section=section)

    def _init_filter(self, filter_):
        name, config = filter_.split(':', 1)
        factory = filter_factories.get(name)
        if factory is None:
            raise ValueError("Unknown filter type: %s" % name)
        return factory(config.strip())

    def reconcile_queues(self, log=None):
        """
        Reconciles queues found in configuration with queues in database.  If
        new queues have been added to the configuration, those queues are
        added to the database.  If old queues have been removed from the
        configuration, they are removed from the database if they are empty.
        If a queue has been removed from the configuration but still has
        queued messages a warning is logged and queue is not removed.
        """
        if log is None:
            log = _NullLog()

        # Reconcile configured queues with queues in db
        configured = self.configured_queues
        with self._get_root() as root:
            # Create new queues
            for queue in configured:
                name = queue['name']
                if name not in root:
                    root[name] = self.Queue()
                    log.info('Created new postoffice queue: %s' % name)

            # Remove old queues if empty
            configured_names = set([q['name'] for q in configured])
            for name, queue in root.items():
                if name not in configured_names:
                    if len(queue):
                        log.warn(
                            "Queue removed from configuration still has "
                            "messages: %s" % name
                        )
                    else:
                        log.info('Removed old postoffice queue: %s' % name)
                        del root[name]

    def import_messages(self, log=None):
        """
        Imports messages from an external maildir, matches them to queues and
        either stores or discards each message depending on whether it matches
        a queue definition.  Once a message is imported it is removed from the
        maildir.
        """
        factory = _message_factory_factory(self, self.MaildirMessage, log)
        maildir = self.Maildir(self.maildir, factory=factory, create=True)
        keys = list(maildir.keys())
        keys.sort()
        for key in keys:
            message = maildir.get_message(key)
            self._import_message(message, log)
            self._archive_message(maildir, message, key)
        n = len(keys)
        if n == 1:
            log.info("Processed one message.")
        else:
            log.info("Processed %d messages." % n)

    def _import_message(self, message, log):
        user = message.get('From')
        if user is None:
            log.info("Message discarded: no 'From' header: %s" %
                     _log_message(message))
            return

        if user == message.get('To'):
            log.info("Message discarded: 'From' and 'To' headers are "
                     "identical: %s" % _log_message(message))
            return

        if not message.get('Message-Id'):
            log.info("Message discarded: no 'Message-Id' header: %s" %
                     _log_message(message))
            return

        if message.get('X-Postoffice') == 'Bounced':
            log.info("Message discarded: ricocheted bounce message: %s" %
                     _log_message(message))
            return

        # Record the message delivery date, in seconds since the epoch,
        # as a header.
        message['X-Postoffice-Date'] = '%d' % message.get_date()

        for filter in self.reject_filters:
            reason = filter(message)
            if reason is not None:
                log.info("Message discarded: rejected by filter: %s: %s" %
                         (reason, _log_message(message)))
                return

        for configured in self.configured_queues:
            filters = configured['filters']
            if not filters or not _filters_match(filters, message):
                continue

            # Matches queue
            with self._get_root() as queues:
                name = configured['name']
                queue = queues[name]

                if queue.is_duplicate(message):
                    log.info("Message discarded: duplicate message: %s" %
                             _log_message(message))
                else:
                    self._check_for_auto_response_and_loops(
                        self, queue, message, log
                    )
                    queue.add(message)
                    queue.collect_frequency_data(message, self.ooo_loop_headers)
                    log.info("Message added to queue, %s: %s" %
                             (name, _log_message(message))
                         )
            break

        else:
            log.info("Message discarded, no matching queues: %s" %
                     _log_message(message))

    def _archive_message(self, maildir, message, key):
        # XXX It would be nice to wire into transaction with a data manager
        today = datetime.date.today().timetuple()[:3]
        name = '%4d.%02d.%02d' % today
        try:
            folder = maildir.get_folder(name)
        except NoSuchMailboxError:
            folder = maildir.add_folder(name)
        folder.add(message)
        maildir.remove(key)

    def _check_for_auto_response_and_loops(self, po, queue, message, log):
        # Like Mailman, if a message has "Precedence: bulk|junk|list",
        # reject it.  The Precedence header is non-standard, yet
        # widely supported.
        precedence = message.get('Precedence', '').lower()
        if precedence in ('bulk', 'junk', 'list'):
            log.info("Message rejected, automatic reply: %s" %
                     _log_message(message))
            message['X-Postoffice-Rejected'] = 'Auto-response'

        # rfc3834 is the standard way to reject automated responses, but
        # it is not yet widely supported.
        auto_submitted = message.get('Auto-Submitted', '').lower()
        if auto_submitted.startswith('auto'):
            log.info("Message rejected, automatic reply: %s" %
                     _log_message(message))
            message['X-Postoffice-Rejected'] = 'Auto-response'

        # Loop Detection
        user = message['From']
        header_tos = message.get_all('To', []) + message.get_all('CC', [])
        orig_to = message.get('X-Original-To')
        if orig_to is None:
            is_bcc = False
        else:
            is_bcc = len([x for x in header_tos if orig_to in x]) == 0
        now = message.get('Date')
        if now is not None:
            now = parsedate(now)
        if now is not None:
            # Certain spambots generate date headers that don't make any
            # sense, eg 32 June, or something crazy like that.
            try:
                now = datetime.datetime(*now[:6])
            except ValueError:
                now = datetime.datetime.now()
        else:
            now = datetime.datetime.now()

        headers = dict([(name, message.get(name))
                        for name in self.ooo_loop_headers])
        freq = self.ooo_loop_frequency
        if queue.is_throttled(user, now, message):
            log.info("Message rejected, user throttled: %s" %
                     _log_message(message))
            message['X-Postoffice-Rejected'] = 'Throttled'

        elif freq and not is_bcc:
            # We don't throttle BCC's because they cannot be loops.
            # The most common case for them is Exchange-driven alias
            # lists.

            # If instanteous or average frequency exceeds limit,
            # throttle user. For average frequency, use interval that
            # is 4 times the inverse of the the freqency. IE, if
            # frequency is 0.25/minute, then 1/frequency is 4 minutes
            # and interval to average over is 16 minutes.
            instant = queue.get_instantaneous_frequency
            average = queue.get_average_frequency
            interval = datetime.timedelta(minutes=4*1/freq)
            if (instant(user, now, headers) > freq or
                average(user, now, interval, headers) > freq):
                queue.throttle(user, now + self.ooo_throttle_period,
                               headers)
                log.info("Message rejected, user triggered "
                         "throttle: %s" % _log_message(message))
                message['X-Postoffice-Rejected'] = 'Throttled'


def _get_opt(config, section, name, default=_marker):
    if config.has_option(section, name):
        return config.get(section, name)
    elif default is not _marker:
        return default
    raise ValueError('Missing required configuration parameter: %s' % name)

def _get_opt_int(config, section, name, default=_marker):
    value = _get_opt(config, section, name, default)
    try:
        return int(value)
    except:
        raise ValueError('Value for %s must be an integer' % name)

def _get_opt_float(config, section, name, default=_marker):
    value = _get_opt(config, section, name, default)
    try:
        return float(value)
    except:
        raise ValueError('Value for %s must be a floating point number' % name)

def _get_opt_list(config, section, name, default=_marker):
    value = _get_opt(config, section, name, default)
    if not value:
        return []
    return [item.strip() for item in value.split(',')]

def _get_opt_bytes(config, section, name, default=_marker):
    value = _get_opt(config, section, name, default).lower()
    num, unit = value, ''
    for i in xrange(len(value)):
        if value[i] < '0' or value[i] > '9':
            num, unit = value[:i], value[i:]
            break
    if not len(num):
        raise ValueError('Value for %s must be an integer' % name)
    value = int(num)
    if not len(unit):
        return value
    elif unit in ('k', 'kb'):
        return value * 1<<10
    elif unit in ('m', 'mb'):
        return value * 1<<20
    elif unit in ('g', 'gb'):
        return value * 1<<30
    else:
        raise ValueError('Bad units in bytes value for %s: %s' % (name, unit))

def _load_fp(fp):
    """
    Read contents of file-like object into memory.
    """
    buf = StringIO()
    shutil.copyfileobj(fp, buf)
    buf.seek(0)
    return buf

def _get_section_indices(fp):
    """
    Get sections in order they appear in a config file.  We do this to
    reconstruct later the order that queues appeared in the config file, since
    ConfigParser does not preserve section ordering.
    """
    indices = {}
    index = 0
    for line in fp:
        if not line.startswith('['):
            continue

        line = line.strip()
        if not line.endswith(']'):
            continue

        indices[line[1:-1]] = index
        index += 1

    return indices

def _filters_match(filters, message):
    for filter_ in filters:
        if not filter_(message):
            return False
    return True

def _ascii_dammit(x):
    if isinstance(x, bytes):
        x = x.decode('ascii', 'replace')
    if isinstance(x, _TEXT):
        x = x.encode('ascii', 'replace')
    return x

def _log_message(message):
    info = ['Message']
    if 'From' in message:
        info.append('From: %s' % _ascii_dammit(message['From']))
    if 'To' in message:
        info.append('To: %s' % _ascii_dammit(message['To']))
    if 'Subject' in message:
        info.append('Subject: %s' % _ascii_dammit(message['Subject']))
    if 'Message-Id' in message:
        info.append('Message-Id: %s' % _ascii_dammit(message['Message-Id']))
    return ' '.join(info)

class _RootContextManagerFactory(object):
    """
    Gets the root postoffice object, an instance of
    `repoze.postoffice.queue.QueuesFolder`.  If folder does not exist it is
    created.  It's parent, however, must already exist if a nested zodb_path
    is used.
    """
    def __init__(self, uri, db_from_uri, path):
        self.uri = uri
        self.db_from_uri = db_from_uri
        self.path = path.strip('/').split('/')

    @contextmanager
    def __call__(self, tm=None):
        if tm is None:
            tm = transaction
        db = self.db_from_uri(self.uri)
        conn = db.open()
        parent = conn.root()
        for name in self.path[:-1]:
            parent = parent[name]
        name = self.path[-1]
        try:
            folder = parent.get(name)
            if folder is None:
                folder = QueuesFolder()
                parent[name] = folder
            yield folder
        except:
            tm.abort()
            raise
        else:
            tm.commit()
        finally:
            conn.close()
            db.close()

def _send_mail(from_addr, to_addrs, message, smtplib=smtplib):
    """
    Sends mail message immediately through SMTP server located on localhost.

    XXX Add some configuration knobs.
    """
    # smtplib passed in for testing
    if not isinstance(message, str):
        message = message.as_string()
    mta = smtplib.SMTP('localhost')
    mta.sendmail(from_addr, to_addrs, message)

def _message_factory_factory(po, wrapped, log):
    def factory(fp):
        # Check size against maximum
        if po.max_message_size:
            fname = fp.name
            if os.path.getsize(fname) > po.max_message_size:
                headers = _read_message_headers(fp)
                log.info("Message rejected, exceeds max size limit: %s"
                         % _log_message(headers))
                message = wrapped()
                for k,v in headers.items():
                    message[k] = v
                message['X-Postoffice-Rejected'] = \
                       'Maximum Message Size Exceeded'
                message.set_payload('Message body discarded.  '
                                    'Maximum message size exceeded.\n\n')
                return message

        return wrapped(fp)

    return factory

_starts_with_whitespace = re.compile('^\s')


def _read_message_headers(fp):
    headers = {}
    header = None
    for line in fp:
        line = line.rstrip('\n').rstrip('\r')
        if not line:
            break
        if _starts_with_whitespace.match(line):
            headers[header] += line
            continue
        header, value = line.split(':', 1)
        headers[header] = value.strip()
    return headers


class _NullLog(object):

    def info(self, *args):
        pass

    def warn(self, *args):
        pass

    def error(self, *args):
        pass
